### ISSUE TITLE
Reliability of glmer to fit Gamma-distributed model - Strange Ranef Variance

### ISSUE THREAD

**Issue by jradinger** [lme4/lme4#643]
I tried to do a simple Gamma distributed GLM (log-link) that also involves random effect and when doing model diagnostics, I realized that the output of glmer() and glmmTMB() are very different in terms of random effects. 

Here an example of what I mean and it's consequences:

```
library(glmmTMB)
library(lme4)

set.seed(123)
N <- 1000
data <- data.frame(y=rgamma(N,3,30),
                   grp=sample(1:20,N,replace=T))
                   
glmmTMB_mod <-glmmTMB(y~1+
                        (1|grp),
                      family=Gamma(link="log"),
                      data=data)

glmer_mod <- glmer(y~1+
                     (1|grp),
                   family=Gamma(link="log"),
                   data=data)
glm_mod <- glm(y~1,
               family=Gamma(link="log"),
               data=data)

fixef(glmer_mod)
fixef(glmmTMB_mod)
VarCorr(glmer_mod)
VarCorr(glmmTMB_mod)

par(mfrow=c(2,2),mar=c(3,3,2,2))
hist(as.matrix(simulate(glmmTMB_mod, nsim = 1))[,1],
     main="glmmTMB",xlab="",breaks=seq(0,1.5,0.05))
hist(as.matrix(simulate(glmer_mod, nsim = 1))[,1],
     main="glmer",xlab="",breaks=seq(0,1.5,0.05))
hist(as.matrix(simulate(glm_mod, nsim = 1))[,1],
     main="glm",xlab="",breaks=seq(0,1.5,0.05))
hist(data$y,main="original distribution",xlab="",breaks=seq(0,1.5,0.05))
```

And here the simulated distributions from the glmmTMB, the glmer, a simplified glm model and the original distribution
![47b58e28-9c74-4fbc-b533-657ebb7cac3f](https://user-images.githubusercontent.com/2982536/135277632-ebe189fe-76c1-4e3b-94b1-6fbf9ae2f73a.png)

Does the lmer model wrongly estimate the random effects in this case(?) Or what is causing the lmer-model to have such shifted simulated distribution compared to the original and the glmmTMB model? Or is this an issue with simulate()?

**Comment by bbolker** [comment]
I'm not sure at the moment. It *may* have to do with the fact that glmmTMB is estimating directly by MLE, while glmer is using the GLM approach of profiling out the dispersion. (I'm not sure, but I think these are guaranteed to be identical for the log link, but not necessarily for the identity link?)  I thought initially it might be a sensitivity to starting conditions, but now I don't think so. I am running some very slow calculations to compare likelihood surfaces, will post when they're done.

**Comment by TiagoAMarques** [comment]
I have found an issue which probably is related to this. I have set up a gitrepos here https://github.com/TiagoAMarques/Report4BB/ which includes inside the folder "testing_lme4" a small script that shows lme4::glmer to be having issues with estimation of parameters for a Gamma GLMM

**Comment by JAQuent** [comment]
Sorry to cross-post here as I am coming from a different angle because I want to fit a BRMS/Stan model.

I also find a strange difference with regard to the random effects between `glmer` and `brm` where the random slopes seem to have much lower variance in the `brm` model. Not sure if these things are related but I thought I mention it here. More details: https://discourse.mc-stan.org/t/honey-i-shrunk-the-random-slopes-too-much/32440

**Comment by Qiwei-Zhao** [comment]
Hi. I met the similar issue when analysis some reaction time data. 

```r

glmer_model <- glmer(RESPONSE_TIME ~ CONDITION + (1|SUBJECT_ID), data = dat, family = Gamma(link = "identity"))
glmmTMB_model <- glmmTMB(RESPONSE_TIME ~ CONDITION + (1|SUBJECT_ID), data = dat, family = Gamma(link = "identity"))
sdat_glmer <- as.matrix(simulate(glmer_model, nsim = 1))
# range(sdat_glmer) # [1] 1.722147e-20 65334.27
sdat_glmer <- sdat_glmer[sdat_glmer <= 5000]
sdat_glmmTMB <- as.matrix(simulate(glmmTMB_model, nsim = 1))

par(mfrow=c(1,3))
hist(dat$RESPONSE_TIME, main = "Original RTs", breaks = seq(0, 5000, 100))
hist(sdat_glmer, main = "Simulated RTs (glmer)", breaks = seq(0, 5000, 100))
hist(sdat_glmmTMB, main = "Simulated RTs (glmmTMB)", breaks = seq(0, 5000, 100))
``` 



<img width="940" alt="image" src="https://github.com/user-attachments/assets/5b8c5d6b-c470-4e36-afb7-766b0fffb737">

**Comment by emjonaitis** [comment]
Add another one to the list. I also noticed nonignorable differences between glmer and glmmTMB when fitting a Gamma model. Using the epil2 dataset that comes with glmmTMB:

```
library(lme4)
library(glmmTMB)
library(dplyr)

data(epil2)
epil2_red <- filter(epil2, y>0)

mod_glmer <- glmer(y ~ trt + (1|subject), data=epil2_red, family=Gamma(link="log"))
mod_glmmTMB <- glmmTMB(y ~ trt + (1|subject), data=epil2_red, family=Gamma(link="log"))

AIC(mod_glmer, mod_glmmTMB)   # very different AIC

summary(mod_glmer)

summary(mod_glmmTMB)
```

There seem to be a couple layers of problems here. Something has changed multiple times with `print` or `summary` for these models, because when I fit with recent archival versions of lme4, the models themselves do not change but the printed variance components do. But the model itself is stably different from the fit in glmmTMB; and for what it's worth, fits in SAS are much closer to glmmTMB. Further, diagnostic plots of the glmer fit look worse than the glmmTMB fit, despite the lower glmer AIC.

**Comment by bbolker** [comment]
I'll take a look. There have been some documented changes (maybe the 2.0-6 change helps things?)

* 2.0-6 (very recently on CRAN): fixed the case where a Gamma GLMM performs poorly for data with small shape parameters by changing the starting parameter values for Gamma families (GH #290)
* 2.0-1:  variances and standard deviations for GLMMs with estimated scale parameters (Gamma etc.) are now printed correctly (the internal estimates were previously, incorrectly, scaled by the estimated dispersion parameters)

**Comment by bbolker** [comment]
I have been exploring this stuff at https://github.com/lme4/lme4/blob/master/misc/GH643.R . It looks like a bit of a can of worms (getting variably different results for `GLMMadaptive`, `glmmTMB`, `glmer`, not entirely sure what to trust/where the differences are coming from.

* several of the problems reported up-thread have been fixed by bugfix in the 2.0-1 release (that was 2026-03-05, so postdating those reports ...) 
* I know that `glmmTMB` uses a different approach to computing the dispersion parameter, but that shouldn't (?) affect the estimation of the random-effects variance
* maybe something to do with starting values? (could try `allFit` ...)

Found out a lot more about this with Claude's help, the required fix is deep, see the `Gamma_GLMM` branch ...

**Comment by bbolker** [comment]
My report on all of this is here ([PDF](https://www.math.mcmaster.ca/bolker/misc/Gamma_GLMM_report.pdf), [qmd](https://github.com/lme4/lme4/blob/Gamma_GLMM/misc/Gamma_GLMM/Gamma_GLMM_report.qmd)), but here is the beginning:

**tl;dr**

We should definitely fix the current implementation of Gamma GLMMs in lme4. Should we
choose an OK, computationally simple method (which I have ready to go), or a better, slightly faster method that would require more architectural changes?

**update**: my current thought is to implement the PIRLS/moment approach in the master branch immediately, and to postpone the joint-phi approach until later (at the very least it would require very careful reverse-dependency checking ...)

## Overview

I’ve discovered some fairly fundamental, longstanding issues with fitting Gamma GLMMs
(more precisely, fitting GLMMs with non-fixed/estimated dispersion parameters, 𝜙) in lme4.
The basic problem is that the PIRLS algorithm we’ve implemented assumes that we can
keep 𝜙 fixed at 1 while fitting the model and estimate it post hoc (e.g. as mean deviance per
observation, or the mean sum of squares of Pearson residuals). This first surfaced as a GitHub
issue, but the real problem was masked for a while by more trivial bugs …

**Comment by ecoRoland2** [comment]
Please ignore if it's not useful. Out of curiosity, I have fit the `schizophrenia` example with brms (using default priors). The fit agrees with the glmmTMB fit.

```r
#lme4 2.0-6, glmmTMB 1.1.14, cmdstanr 0.9.0
library(lme4)

data(schizophrenia)
schizophrenia$id <- factor(schizophrenia$id)

fit1 <- glmer(imps79 ~ TxDrug * Week + (1 | id), data = schizophrenia, family = Gamma(link = "log"))

library(glmmTMB)
fit2 <- glmmTMB(imps79 ~ TxDrug * Week + (1 | id), data = schizophrenia, family = Gamma(link = "log"))

library(brms)
#install.packages("cmdstanr", repos = c('https://stan-dev.r-universe.dev', getOption("repos")))

default_prior(imps79 ~ TxDrug * Week + (1 | id), data = schizophrenia, family = Gamma(link = "log"))
 #                  prior     class        coef group resp dpar nlpar lb ub tag       source
 #                 (flat)         b                                                  default
 #                 (flat)         b      TxDrug                                 (vectorized)
 #                 (flat)         b TxDrug:Week                                 (vectorized)
 #                 (flat)         b        Week                                 (vectorized)
 # student_t(3, 1.5, 2.5) Intercept                                                  default
 #   student_t(3, 0, 2.5)        sd                                    0             default
 #   student_t(3, 0, 2.5)        sd                id                  0        (vectorized)
 #   student_t(3, 0, 2.5)        sd   Intercept    id                  0        (vectorized)
 #      gamma(0.01, 0.01)     shape                                    0             default

fit3 <- brm(imps79 ~ TxDrug * Week + (1 | id), data = schizophrenia, family = Gamma(link = "log"),
            backend = "cmdstanr",
            cores = 4, seed = 42)

VarCorr(fit1)
#Groups Name        Std.Dev.
#id     (Intercept) 0.66133
VarCorr(fit2)
#Conditional model:
# Groups Name        Std.Dev.
# id     (Intercept) 0.19831 
VarCorr(fit3)
#$id
#$id$sd
#           Estimate Est.Error      Q2.5     Q97.5
#Intercept 0.1994632 0.0112946 0.1776732 0.2221187

fixef(fit1)
#(Intercept)      TxDrug        Week TxDrug:Week 
#1.64951500 -0.03590826 -0.03707484 -0.06695051 
fixef(fit2)
#Conditional model:
#(Intercept)       TxDrug         Week  TxDrug:Week  
#    1.66202     -0.03405     -0.03707     -0.06334
fixef(fit3)
#               Estimate   Est.Error        Q2.5       Q97.5
#Intercept    1.66098390 0.029072414  1.60343528  1.71837475
#TxDrug      -0.03222311 0.032982802 -0.09758523  0.03362996
#Week        -0.03689579 0.007004922 -0.05062067 -0.02322655
#TxDrug:Week -0.06359483 0.007879594 -0.07912515 -0.04810335

sigma(fit1)
#[1] 0.2754139
sigma(fit2)
#[1] 0.2888543
mean(1/sqrt(as_draws_array(fit3, variable = "shape")))
#[1] 0.2893272


```

**Comment by bbolker** [comment]
Interesting but not surprising. I didn't bother to include Bayesian approaches in the exploration because the differences (priors, estimating marginal means vs modes/MLEs) would add another layer of complication ... especially when the differences are small, as in this case

**Comment by bbolker** [comment]
fixed by Gamma_GLMM merge
